### PR TITLE
Che on Windows - copy ws-agent from che.properties location.

### PR DESCRIPTION
**Pull request by vparfonov**
Copy terminal,ws-agent and ext-config to the Windows User Home need for mountig this directory to the docker container

### CODE REVIEW

**Comment by vparfonov** [comment]
@skabashnyuk @sleshchenko @garagatyi

File: plugin-docker/che-plugin-docker-machine/src/main/java/org/eclipse/che/plugin/docker/machine/WindowsHostUtils.java [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+ *******************************************************************************/
+package org.eclipse.che.plugin.docker.machine;
+
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+
+/**
+ * //
```
Comment (garagatyi) [inline plugin-docker/che-plugin-docker-machine/src/main/java/org/eclipse/che/plugin/docker/machine/WindowsHostUtils.java]: Please explain purpose of this class in javadocs

File: plugin-docker/che-plugin-docker-machine/src/main/java/org/eclipse/che/plugin/docker/machine/WindowsHostUtils.java [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+ *
+ * @author Vitalii Parfonov
+ */
+public class WindowsHostUtils {
+
+    /**
+     * Path to the Che configuration folder on Windows in user home.
+     * According to some limitations of docker+Windows all monted to the docker container folder must be in user.home
+     * directory.
+     *
```
Comment (garagatyi) [inline plugin-docker/che-plugin-docker-machine/src/main/java/org/eclipse/che/plugin/docker/machine/WindowsHostUtils.java]: extra line

File: plugin-docker/che-plugin-docker-machine/src/main/java/org/eclipse/che/plugin/docker/machine/WindowsHostUtils.java [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+     */
+    public static final String WINDOWS_CHE_HOME_PATH = "\\AppData\\Local\\Eclipse Che";
+
+    /**
+     * Create Che Home directory if not exist on Windows hosts it will be locate in
+     * System.getProperty("user.home") + {@link WindowsHostUtils#WINDOWS_CHE_HOME_PATH}
+     * @return Path to Che home dir
```
Comment (garagatyi) [inline plugin-docker/che-plugin-docker-machine/src/main/java/org/eclipse/che/plugin/docker/machine/WindowsHostUtils.java]: add new line

File: plugin-java/che-plugin-java-ext-debugger-java-client/src/main/java/org/eclipse/che/ide/ext/java/jdi/client/actions/ServerLogAction.java [hunk 4]
```diff
@@ -1,88 +0,0 @@
-/*******************************************************************************
```
Comment (garagatyi) [inline plugin-java/che-plugin-java-ext-debugger-java-client/src/main/java/org/eclipse/che/ide/ext/java/jdi/client/actions/ServerLogAction.java]: Why you removed this file in this PR?

File: plugin-docker/che-plugin-docker-machine/src/main/java/org/eclipse/che/plugin/docker/machine/ext/provider/TerminalVolumeProvider.java [hunk 5]
```diff
@@ -33,9 +45,17 @@
     @Override
     public String get() {
         if (SystemInfo.isWindows()) {
-            return System.getProperty("user.home") + "\\AppData\\Local\\che\\terminal" + ":/mnt/che/terminal:ro";
+            try {
+                final Path cheHome = WindowsHostUtils.ensureCheHomeExist();
+                final Path terminalPath = cheHome.resolve(TERMINAL);
+                IoUtil.copy(Paths.get(terminalArchivePath).toFile(), terminalPath.toFile(), null, true);
```
Comment (garagatyi) [inline plugin-docker/che-plugin-docker-machine/src/main/java/org/eclipse/che/plugin/docker/machine/ext/provider/TerminalVolumeProvider.java]: Why you use different APIs for copying files in terminal and wsagent providers?

Comment (vparfonov) [inline plugin-docker/che-plugin-docker-machine/src/main/java/org/eclipse/che/plugin/docker/machine/ext/provider/TerminalVolumeProvider.java]: IoUtil.copy better solution for copy directories

File: plugin-java/che-plugin-java-ext-debugger-java-client/src/main/java/org/eclipse/che/ide/ext/java/jdi/client/actions/ServerLogAction.java (hunk 4 above)
Comment (vparfonov) [inline plugin-java/che-plugin-java-ext-debugger-java-client/src/main/java/org/eclipse/che/ide/ext/java/jdi/client/actions/ServerLogAction.java]: my fail

**Comment by garagatyi** [comment]
ok

File: plugin-docker/che-plugin-docker-machine/src/main/java/org/eclipse/che/plugin/docker/machine/ext/provider/ExtServerVolumeProvider.java [hunk 6]
```diff
@@ -11,11 +11,18 @@
...
+import org.eclipse.che.plugin.docker.machine.WindowsHostUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
 
 import javax.inject.Inject;
 import javax.inject.Named;
 import javax.inject.Provider;
 import javax.inject.Singleton;
+import java.io.IOException;
+import java.nio.file.*;
```
Comment (sleshchenko) [inline plugin-docker/che-plugin-docker-machine/src/main/java/org/eclipse/che/plugin/docker/machine/ext/provider/ExtServerVolumeProvider.java]: We should not use imports with *

File: plugin-docker/che-plugin-docker-machine/src/main/java/org/eclipse/che/plugin/docker/machine/ext/provider/DockerExtConfBindingProvider.java [hunk 7]
```diff
@@ -27,22 +33,38 @@
  */
 @Singleton
 public class DockerExtConfBindingProvider implements Provider<String> {
-    public static final  String EXT_CHE_LOCAL_CONF_DIR = "/mnt/che/conf";
-    private static final String CONTAINER_TARGET       = ":" + EXT_CHE_LOCAL_CONF_DIR + ":ro";
+
+    public static final String EXT_CHE_LOCAL_CONF_DIR = "/mnt/che/conf";
+
+    private static final String PLUGIN_CONF      = "plugin-conf";
```
Comment (sleshchenko) [inline plugin-docker/che-plugin-docker-machine/src/main/java/org/eclipse/che/plugin/docker/machine/ext/provider/DockerExtConfBindingProvider.java]: It makes sense to have the same names for folders with configuration of workspace agent. Your proposal to use ext and plugin-conf. Why?

**Comment by sleshchenko** [comment]
other OK for me

File: plugin-docker/che-plugin-docker-machine/src/main/java/org/eclipse/che/plugin/docker/machine/ext/provider/DockerExtConfBindingProvider.java (hunk 7 above)
Comment (vparfonov) [inline plugin-docker/che-plugin-docker-machine/src/main/java/org/eclipse/che/plugin/docker/machine/ext/provider/DockerExtConfBindingProvider.java]: yes i are right better use equals names

**Pull request merged**